parser-typechecker/src/Unison/Util/Rope.hs

```diff
@@ -0,0 +1,197 @@
+{-# Language FunctionalDependencies #-}
+{-# Language DeriveFoldable #-}
+
+module Unison.Util.Rope 
+  (chunks, singleton, one, map, traverse, null, flatten, two, cons, uncons, snoc, unsnoc, 
+   Sized(..), Take(..), Drop(..), Reverse(..), Index(..), Rope,
+   ) 
+where
+
+import Prelude hiding (drop,take,reverse,map,traverse,null)
+import Data.Foldable (toList)
+import Control.DeepSeq (NFData(..))
+
+-- | Size-balanced binary tree of chunks. 
+-- The `Int` in the `Two` constructor is a cached size of that subtree.
+data Rope a
+  = Empty
+  | One {-# unpack #-} !a
+  | Two {-# unpack #-} !Int !(Rope a) !(Rope a)
+  deriving Foldable
+
+chunks :: Rope a -> [a]
+chunks = toList
+
+singleton, one :: Sized a => a -> Rope a
+one a | size a == 0 = Empty
+one a = One a
+singleton = one
+
+map :: Sized b => (a -> b) -> Rope a -> Rope b
```

Comment:
`map` and `traverse` don't seem to re-balance afterwards.
If there's an invariant that the caller of map shouldn't change the size of the elements during mapping/traversing then that should probably be indicated; especially since a type-changing update could very easily do so.

It'd also be worth noting that the `a`'s that are being mapped will be chunked at random, which means certain maps aren't valid, e.g. `Text.words <$> myRope` would give some unexpected results if chunk-splits happen in the middle of words, this is expected of course if you understand how Ropes work.

Reply:
Yeah, I'll add a note. It's often okay to change the size of chunks. It might lead to a less size-balanced tree, but that's often fine. Like going from `Rope Text` to `Rope ByteString` by utf8 encoding each chunk is fine and doesn't need rebalancing

Mapping (say) 95% of the chunks to empty is where you want to rebalance.